### ISSUE TITLE
Confusing error messages when expressions contain missing assets

### ISSUE THREAD

**Issue by mrob95** [cogeotiff/rio-tiler#620]
Repro:
```python
from rio_tiler.io.stac import STACReader


r = STACReader(
    input="https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/48/Q/UM/2021/3/S2B_48QUM_20210312_0_L2A/S2B_48QUM_20210312_0_L2A.json",
)

# # Non-existent asset:
# # rio_tiler.errors.MissingAssets: assets must be passed either via expression or assets options.
t = r.tile(0, 0, 0, expression="abc-0.1", asset_as_band=True)

# # Non-existent asset with assets passed explicitly:
# # rio_tiler.errors.MissingAssets: assets must be passed either via expression or assets options.
# t = r.tile(0, 0, 0, expression="abc-0.1", assets=["abc"], asset_as_band=True)

# # One non-existent asset, one valid asset:
# # KeyError: 'abc'
# t = r.tile(0, 0, 0, expression="blue+abc", asset_as_band=True)

# # Valid asset without asset_as_band set
# # rio_tiler.errors.MissingAssets: assets must be passed either via expression or assets options.
# t = r.tile(0, 0, 0, expression="blue+0.1")

print(t)
```

When calling `tile` on a `STACReader` with an expression that contains assets/bands that don't exist (in the above case 'blue' exists but 'abc' doesn't), the error message doesn't give much indication about what the actual problem is. 

It would be great if the error message said something like "asset 'abc' was not found in the STAC item, which contains assets a, b, c, ..." or similar.

I also wonder whether, when asset_as_band=False and no bands are found in the expression, `rio-tiler` should try to parse the expression again with asset_as_band=True and if assets are found, report an error like "items in expression were not matched as bands but could be matched as asset, maybe you meant to pass asset_as_band=True?". This change confused me for a while when updating from an older version.

Thanks
Mike

**Comment by vincentsarago** [comment]
Thanks for opening this issue @mrob95 

I'll add a fix for most of ☝️ 

The only one which is tricky is for the `KeyError` because this happens in numexpr

**Comment by mrob95** [comment]
Yeah, that is probably the least bad one anyway since it does kind of indicate what the problem was. I wonder though whether that would cause a 500 in titiler? Not at my laptop to test atm

**Comment by vincentsarago** [comment]
by default it would cause a 500 

we only catch some of the error https://github.com/developmentseed/titiler/blob/d405395e2047629c8e35168ac57c0e83f5b8943d/src/titiler/core/titiler/core/errors.py#L33-L46

**Comment by vincentsarago** [comment]
Fix published in 5.0.1 and 4.1.13 🥳

**Comment by mrob95** [comment]
Thanks for the fix, and for the great libraries 😃
